agent/service.go

```diff
@@ -15,165 +15,193 @@ package main
 	}
-	listener, err := vsock.Listen(port)
+
+	// Override the incoming stdio FIFOs, which have paths from the host that we can't use
+	fifoSet, err := cio.NewFIFOSetInDir(bundleDir.RootPath(), req.ID, req.Terminal)
 	if err != nil {
-		log.G(ctx).WithError(err).Error("unable to listen on vsock")
-		f.Close()
-		return
+		log.G(ctx).WithError(err).Error("failed opening stdio FIFOs")
+		return nil, err
 	}
 
-	var conn net.Conn
-	for {
-		// accept is non-blocking so try to accept until we get
-		// a connection
-		// TODO: investigate if there is a way to distinguish
-		// transient errors from permanent ones.
-		conn, err = listener.Accept()
-		if err != nil {
-			continue
-		}
-		break
+	// Don't try to connect any io streams that weren't requested by the client
+	if req.Stdin == "" {
+		fifoSet.Stdin = ""
 	}
-	go func() {
-		<-ctx.Done()
-		conn.Close()
-		f.Close()
-	}()
-	log.G(ctx).Debug("begin copying io")
-	buf := make([]byte, internal.DefaultBufferSize)
-	if in {
-		_, err = io.CopyBuffer(f, conn, buf)
-	} else {
-		_, err = io.CopyBuffer(conn, f, buf)
+
+	if req.Stdout == "" {
+		fifoSet.Stdout = ""
 	}
+
+	if req.Stderr == "" {
+		fifoSet.Stderr = ""
+	}
+
+	task, err := ts.AddTask(ctx, req.ID, runcService, bundleDir, extraData, fifoSet)
 	if err != nil {
-		log.G(ctx).WithError(err).Error("error with stdio")
+		return nil, err
 	}
-}
 
-func unpackBundle(path string, bundle *types.Any) (*types.Any, error) {
-	// get json bytes from task request
-	extraData := &proto.ExtraData{}
-	err := types.UnmarshalAny(bundle, extraData)
+	logger.Debug("calling runc create")
+
+	// Override some of the incoming paths, which were set on the Host and thus not valid here in the Guest
+	req.Bundle = bundleDir.RootPath()
+	req.Rootfs = nil
+	req.Stdin = fifoSet.Stdin
+	req.Stdout = fifoSet.Stdout
+	req.Stderr = fifoSet.Stderr
+
+	// Just provide runc the options it knows about, not our wrapper
+	req.Options = task.ExtraData().GetRuncOptions()
+
+	// Start the io proxy and wait for initialization to complete before starting
+	// the task to ensure we capture all task output
+	err = <-task.StartStdioProxy(ctx, vm.VSockToFIFO, acceptVSock)
 	if err != nil {
 		return nil, err
 	}
-	// write bundle/config.json bytes
-	err = ioutil.WriteFile(path, extraData.JsonSpec, 0644)
+
+	resp, err := task.Create(ctx, req)
 	if err != nil {
+		logger.WithError(err).Error("error creating container")
 		return nil, err
 	}
-	return extraData.RuncOptions, nil
+
+	logger.WithField("pid", resp.Pid).Debugf("create succeeded")
+	return resp, nil
+}
+
+var _ vm.VSockConnector = acceptVSock
+
+func acceptVSock(ctx context.Context, port uint32) (net.Conn, error) {
+	listener, err := vsock.Listen(port)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to listen on vsock")
+	}
+
+	var conn net.Conn
+	for range time.Tick(10 * time.Millisecond) {
+		// accept is non-blocking so try to accept until we get a connection
+		// TODO: investigate if there is a way to distinguish
```

Comment:
This might help: https://github.com/containerd/ttrpc/blob/master/server.go#L94

Reply:
This comment was carried over from the previous implementation. I'll take a look at that though either way, thanks. If it's a one line change I'll include it in this PR, but otherwise I'll save it for a next iteration.

Reply:
It was pretty straightforward, updated with that as a separate commit